p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,230 @@
+  // There is always a most general match because of P4 restrications
+  // on match combination (which may be under-documented).
+  // See https://github.com/p4lang/p4-spec/issues/411
+  p4_symbolic.bmv2.TableMatchTypeEnum match_type = 5;
+  // Table type with respect to action selection.
+  p4_symbolic.bmv2.TableTypeEnum type = 6;
+}
+
+// An abstract p4 statement corresponding to a top level operation within
+// an action body.
+message Statement {
+  p4_symbolic.bmv2.SourceLocation source_info = 1;
+  // Various statement concrete types.
+  oneof statement {
+    AssignmentStatement assignment_statement = 2;
+  }
+}
+
+// An assignment statement of the form <left> = <right>.
+// Note that left can be an rvalue expression.
+message AssignmentStatement {
+  LValue left = 1;
+  RValue right = 2;
+}
+
+// The structure of an abstract RValue.
+message RValue {
+  oneof rvalue {
+    FieldValue field_value = 2;
+    HexstrValue hexstr_value = 3;
+    BoolValue bool_value = 4;
+    StringValue string_value = 5;
+    Variable variable_value = 6;
+    RExpression expression_value = 7;
+  }
+}
+
+// The structure of an abstract LValue.
+message LValue {
+  oneof lvalue {
+    FieldValue field_value = 1;
+    Variable variable_value = 2;
+  }
+}
+
+// A header field accessed statically.
+// Field access that is dynamic (either in the index/offset or in the header
+// object) is defined under AccessExpression below.
+message FieldValue {
+  string header_name = 1;
+  string field_name = 2;
+}
+
+// Corresponds to a hex bigendiean encoding of a bit-string.
+// In the bmv2 JSON format. The hexstring can be prefixed by "-"
+// for negative values.
+// We remove any such prefix and explicitly set the negative flag.
+message HexstrValue {
+  string value = 1;
+  bool negative = 2;
+}
+
+// A boolean literal value.
+message BoolValue {
+  bool value = 1;
+}
+
+// A string literal value.
+message StringValue {
+  string value = 1;
+}
+
+// Our explicit representation of a runtime variable.
+// The bmv2 JSON format represents variables as indices
+// to a runtime_data array in the enclosing scope, which
+// contains the variable name.
+// We directly refer to variables by their name.
+message Variable {
+  string name = 1;
+}
+
+// An abstract RExpression.
+message RExpression {
+  oneof expression {
+    BinaryExpression simple_binary_expression = 1;
+    UnaryExpression simple_unary_expression = 2;
+    TernaryExpression ternary_expression = 3;
+  }
+}
+
+// Arithmetic, relational, and boolean binary operations.
+message BinaryExpression {
+  // Supported basic binary expressions.
+  enum Operation {
+    plus = 0;
+    minus = 1;
+    times = 3;
+    left_shift = 4;   // <<
+    right_shift = 5;  // >>
+    equals = 6;       // ==
+    not_equals = 7;
+    greater = 8;
+    greater_equals = 9;
+    less = 10;
+    less_equals = 11;
+    and = 12;
+    or = 13;
+    bit_and = 14;  // &
+    bit_or = 15;
+    bit_xor = 16;
+  }
+
+  Operation op = 1;
```

Comment:
I would really prefer operation over op here, especially since we have both operation and operand below.

Reply:
yeah makes sense. These are remnants of the JSON format, but we do not need to stick to it in the IR